src/mapreduce.jl

```diff
@@ -1,430 +1,214 @@
-        newtype = similar_type(a, Size(Snew))
-
-        exprs = Array{Expr}(Snew)
-        itr = [1:n for n = Snew]
-        for i = Base.product(itr...)
-            ik = copy([i...])
-            ik[D] = 2
-            expr = :(op(a[$(i...)], a[$(ik...)]))
-            for k = 3:S[D]
-                ik[D] = k
-                expr = :(op($expr, a[$(ik...)]))
-            end
-
-            exprs[i...] = expr
-        end
-
-        return quote
-            $(Expr(:meta,:inline))
-            @inbounds return $(Expr(:call, newtype, Expr(:tuple, exprs...)))
-        end
-    end
+@inline function map!(f, dest::StaticArray, a::StaticArray, b::StaticArray)
+    _map!(f, dest, same_size(dest, a, b), a, b)
 end
 
-# These are all similar in Base but not @inline'd
-@inline sum{T}(a::StaticArray{T}) = reduce(+, zero(T), a)
-@inline prod{T}(a::StaticArray{T}) = reduce(*, one(T), a)
-@inline count(a::StaticArray{Bool}) = reduce(+, 0, a)
-@inline all(a::StaticArray{Bool}) = reduce(&, true, a)  # non-branching versions
-@inline any(a::StaticArray{Bool}) = reduce(|, false, a) # (benchmarking needed)
-@inline mean(a::StaticArray) = sum(a) / length(a)
-@inline sumabs{T}(a::StaticArray{T}) = mapreduce(abs, +, zero(T), a)
-@inline sumabs2{T}(a::StaticArray{T}) = mapreduce(abs2, +, zero(T), a)
-@inline minimum(a::StaticArray) = reduce(min, a) # base has mapreduce(idenity, scalarmin, a)
-@inline maximum(a::StaticArray) = reduce(max, a) # base has mapreduce(idenity, scalarmax, a)
-@inline minimum{D}(a::StaticArray, dim::Type{Val{D}}) = reducedim(min, a, dim)
-@inline maximum{D}(a::StaticArray, dim::Type{Val{D}}) = reducedim(max, a, dim)
-
-@generated function diff{D}(a::StaticArray, ::Type{Val{D}}=Val{1})
-    S = size(a)
-    N = ndims(a)
-    Snew = ([n==D ? S[n]-1 : S[n] for n = 1:N]...)
-    newtype = similar_type(a, Size(Snew))
 
-    exprs = Array{Expr}(Snew)
-    itr = [1:n for n = Snew]
-
-    for i1 = Base.product(itr...)
-        i2 = copy([i1...])
-        i2[D] = i1[D] + 1
-        exprs[i1...] = :(a[$(i2...)] - a[$(i1...)])
+@generated function _map!(f, dest, ::Size{S}, a::StaticArray...) where {S}
+    exprs = Vector{Expr}(prod(S))
+    for i ∈ 1:prod(S)
+        tmp = [:(a[$j][$i]) for j ∈ 1:length(a)]
+        exprs[i] = :(dest[$i] = f($(tmp...)))
     end
-
     return quote
-        $(Expr(:meta,:inline))
-        @inbounds return $(Expr(:call, newtype, Expr(:tuple, exprs...)))
+        @_inline_meta
+        @inbounds $(Expr(:block, exprs...))
     end
 end
 
 ###############
 ## mapreduce ##
 ###############
 
-# Single array
-@generated function mapreduce(f, op, a1::StaticArray)
-    if length(a1) == 1
-        return :(f(a1[1]))
-    else
-        expr = :(op(f(a1[1]), f(a1[2])))
-        for j = 3:length(a1)
-            expr = :(op($expr, f(a1[$j])))
-        end
-        return quote
-            $(Expr(:meta, :inline))
-            @inbounds return $expr
-        end
-    end
+@inline function mapreduce(f, op, a::StaticArray...)
```

Comment:
Just trying to think of ways in which the 0-occurrences version of this might be a problem (or not). I guess there could also be other packages that might do the same basic signature but for vararg of their own type, in which case `mapreduce(f, op)` might be ambiguous between here and that other package? Maybe a little safer to have the signature 1-or-more instead of 0-or-more, but maybe I'm just being paranoid.